scrapy/core/http2/protocol.py

```diff
@@ -0,0 +1,191 @@
+        self._metadata = {
+            'certificate': None,
+            'ip_address': None
+        }
+
+    def _stream_close_cb(self, stream_id: int):
+        """Called when stream is closed completely
+        """
+        self.streams.pop(stream_id, None)
+
+    def _new_stream(self, request: Request):
+        """Instantiates a new Stream object
+        """
+        stream_id = next(self.stream_id_count)
+
+        stream = Stream(
+            stream_id=stream_id,
+            request=request,
+            connection=self.conn,
+            conn_metadata=self._metadata,
+            write_to_transport=self._write_to_transport,
+            cb_close=self._stream_close_cb
+        )
+
+        self.streams[stream.stream_id] = stream
+        return stream
+
+    def _send_pending_requests(self):
+        # TODO: handle MAX_CONCURRENT_STREAMS
+        # Initiate all pending requests
+        while self._pending_request_stream_pool:
+            stream = self._pending_request_stream_pool.popleft()
+            stream.initiate_request()
+
+    def _write_to_transport(self):
+        """ Write data to the underlying transport connection
+        from the HTTP2 connection instance if any
+        """
+        data = self.conn.data_to_send()
+        self.transport.write(data)
+
+        LOGGER.debug("Sent {} bytes to {} via transport".format(len(data), self._metadata['ip_address']))
+
+    def request(self, _request: Request):
+        stream = self._new_stream(_request)
+        d = stream.get_response()
+
+        # If connection is not yet established then add the
+        # stream to pool or initiate request
+        if self.is_connection_made:
+            stream.initiate_request()
+        else:
+            self._pending_request_stream_pool.append(stream)
+
+        return d
+
+    def connectionMade(self):
+        """Called by Twisted when the connection is established. We can start
+        sending some data now: we should open with the connection preamble.
+        """
+        self.destination = self.transport.getPeer()
+        LOGGER.info('Connection made to {}'.format(self.destination))
+
+        self._metadata['certificate'] = Certificate(self.transport.getPeerCertificate())
+        self._metadata['ip_address'] = ipaddress.ip_address(self.destination.host)
+
+        self.conn.initiate_connection()
+        self._write_to_transport()
+        self.is_connection_made = True
+
+    def dataReceived(self, data):
+        events = self.conn.receive_data(data)
+        self._handle_events(events)
+        self._write_to_transport()
+
+    def connectionLost(self, reason=connectionDone):
+        """Called by Twisted when the transport connection is lost.
+        No need to write anything to transport here.
+        """
+        # Pop all streams which were pending and were not yet started
+        for stream_id in list(self.streams):
+            self.streams[stream_id].close(StreamCloseReason.CONNECTION_LOST, reason)
```

Comment:
```suggestion
        for stream in self.streams.values():
            stream.close(StreamCloseReason.CONNECTION_LOST, reason)
```
Would "close" be more accurate than "pop" in the above comment?

Reply:
> Would "close" be more accurate than "pop" in the above comment?

I did not get you?

Reply:
@elacuesta Some background: https://github.com/scrapy/scrapy/pull/4610#discussion_r440006977

Reply:
Oh right, I remember that. Makes sense, and the comment in https://github.com/scrapy/scrapy/commit/23906b6bee953d9bc5dd8042e785711b11840797#diff-a9927075d70e58a912f1e78ff3251186R143-R145 helps, but I'm not sure I like it. I will suggest a change in that thread.